File: typescript/src/ai/embeddings/iEmbeddingIndex.ts
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { Embedding } from './embedding';
import { IEmbeddingWithMetadata } from './iEmbeddingWithMetadata';

/**
 * Represents an searchable index of Embedding<TEmbedding> structs.
 */
export interface IEmbeddingIndex {
    /**
     * Gets the nearest matches to the Embedding<TEmbedding>.
     * @param container The storage container to search.
     * @param embedding The input Embedding<TEmbedding> to use as the search.
     * @param topNCount The max number of results to return.
     * @param minScore The minimum score to consider in the distance calculation.
     * @returns A tuple consisting of the IEmbeddingWithMetadata<TEmbedding> and the similarity score as a double.
     */
    getNearestNMatchesAsync(
        container: string,
        embedding: Embedding,
        topNCount?: number,
        minScore?: number
    ): Promise<[IEmbeddingWithMetadata, number]>;
}

/**
 * Common extension methods for IEmbeddingIndex<TEmbedding> objects.
 */
export class EmbeddingIndexExtensions {
    public static getNearestMatchAsync(
        index: IEmbeddingIndex,
        container: string,
        embedding: Embedding,
        minScore?: number
    ): Promise<[IEmbeddingWithMetadata, number]> {
        if (!index) {
            throw new Error('Embedding index cannot be NULL');
        }

        return index.getNearestNMatchesAsync(container, embedding, 1, minScore).then((match) => match || [null, 0]);
    }
}

File: dotnet/src/SemanticKernel/AI/Embeddings/IEmbeddingIndex.cs
// Copyright (c) Microsoft. All rights reserved.

using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.SemanticKernel.Diagnostics;

namespace Microsoft.SemanticKernel.AI.Embeddings;

/// <summary>
/// Represents an searchable index of <see cref="Embedding{TEmbedding}"/> structs.
/// </summary>
/// <typeparam name="TEmbedding">The data type of the embedding.</typeparam>
public interface IEmbeddingIndex<TEmbedding>
    where TEmbedding : unmanaged
{
    /// <summary>
    /// Gets the nearest matches to the <see cref="Embedding{TEmbedding}"/>.
    /// </summary>
    /// <param name="collection">The storage collection to search.</param>
    /// <param name="embedding">The input <see cref="Embedding{TEmbedding}"/> to use as the search.</param>
    /// <param name="limit">The max number of results to return.</param>
    /// <param name="minRelevanceScore">The minimum score to consider in the distance calculation.</param>
    /// <returns>A tuple consisting of the <see cref="IEmbeddingWithMetadata{TEmbedding}"/> and the similarity score as a <see cref="double"/>.</returns>
    IAsyncEnumerable<(IEmbeddingWithMetadata<TEmbedding>, double)> GetNearestMatchesAsync(
        string collection,
        Embedding<TEmbedding> embedding,
        int limit = 1,
        double minRelevanceScore = 0.0);
}

/// <summary>
/// Common extension methods for <see cref="IEmbeddingIndex{TEmbedding}"/> objects.
/// </summary>
public static class EmbeddingIndexExtensions
{
    /// <summary>
    /// Searches the index for the nearest match to the <see cref="Embedding{TEmbedding}"/>.
    /// </summary>
    public static async Task<(IEmbeddingWithMetadata<TEmbedding>, double)> GetNearestMatchAsync<TEmbedding>(this IEmbeddingIndex<TEmbedding> index,
        string collection,
        Embedding<TEmbedding> embedding,
        double minScore = 0.0)
        where TEmbedding : unmanaged
    {
        Verify.NotNull(index);
        await foreach (var match in index.GetNearestMatchesAsync(collection, embedding, 1, minScore))
        {
            return match;
        }

        return default;
    }
}
